contracts/ENSGovernorV2.sol

```diff
@@ -0,0 +1,187 @@
+
+    /**
+     * @dev Emitted when a bond is created.
+     * @param proposer The address of the proposer.
+     * @param amount The amount of bond.
+     */
+    event BondCreated(address indexed proposer, uint256 amount);
+
+    constructor(
+        ENSToken _token,
+        TimelockController _timelock
+    ) ENSGovernor(_token, _timelock) {}
+
+    /**
+     * @dev Propose a governance action with a bond.
+     * @param targets The addresses to call.
+     * @param values The values to send.
+     * @param calldatas The calldata to send.
+     * @param description The description of the proposal.
+     * @return proposalId The ID of the proposal.
+     */
+    function proposeWithBond(
+        address[] memory targets,
+        uint256[] memory values,
+        bytes[] memory calldatas,
+        string memory description
+    ) public returns (uint256 proposalId) {
+        // Transfer bond to the contract
+        uint256 bondAmount = calculateBond(targets.length);
+        require(
+            token.transferFrom(msg.sender, address(this), bondAmount),
+            "ERC20: transferFrom failed"
+        );
+        lockedBondsBalance += bondAmount;
+
+        // Create the proposal
+        proposalId = propose(targets, values, calldatas, description);
+
+        // Store the bond details
+        _proposalBonds[proposalId] = ProposalBond({
+            amount: bondAmount,
+            refunded: false,
+            proposer: msg.sender
+        });
+
+        emit BondCreated(msg.sender, bondAmount);
+    }
+
+    /**
+     * @dev Calculate the bond amount based on the count of targets.
+     * @param countOfTargets The number of targets in the proposal.
+     * @return The calculated bond amount.
+     */
+    function calculateBond(
+        uint256 countOfTargets
+    ) public view returns (uint256) {
+        return countOfTargets * ensPerTarget;
+    }
+
+    /**
+     * @dev Cancel a proposal and refund the bond if necessary.
+     * @param targets The addresses to call.
+     * @param values The values to send.
+     * @param calldatas The calldata to send.
+     * @param descriptionHash The hash of the proposal description.
+     * @return proposalId The ID of the proposal.
+     */
+    function _cancel(
+        address[] memory targets,
+        uint256[] memory values,
+        bytes[] memory calldatas,
+        bytes32 descriptionHash
+    ) internal override(ENSGovernor) returns (uint256) {
+        uint256 proposalId = super._cancel(
+            targets,
+            values,
+            calldatas,
+            descriptionHash
+        );
+        (
+            uint256 againstVotesWithoutBond,
+            uint256 againstVotes, // votesLength
+            ,
+
+        ) = proposalVotes(proposalId);
+
+        ProposalBond storage bond = _proposalBonds[proposalId];
+
+        if (againstVotes >= againstVotesWithoutBond) {
+            require(
+                token.transfer(bond.proposer, bond.amount),
```

Comment:
use `safeTransferFrom` from the treasury, as the other comment.

Reply:
Done